creusot/src/clone_map.rs

```diff
@@ -48,72 +54,61 @@ impl PreludeModule {
     }
 }
 
+#[derive(Clone)]
 pub struct CloneMap<'tcx> {
     tcx: TyCtxt<'tcx>,
     prelude: IndexSet<PreludeModule>,
     names: IndexMap<(DefId, SubstsRef<'tcx>), CloneInfo>,
     count: usize,
+    item_type: ItemType,
 }
 
 #[derive(Clone)]
-struct CloneInfo {
+pub struct CloneInfo {
     name: String,
-    projs: Vec<CloneSubst>,
     hidden: bool,
 }
 
 impl CloneInfo {
     fn from_name(name: String) -> Self {
-        CloneInfo { name, projs: Vec::new(), hidden: false }
+        CloneInfo { name, hidden: false }
     }
 
     fn hidden(name: String) -> Self {
-        CloneInfo { name, projs: Vec::new(), hidden: true }
+        CloneInfo { name, hidden: true }
     }
-}
 
-impl<'tcx> CloneMap<'tcx> {
-    pub fn new(tcx: TyCtxt<'tcx>) -> Self {
-        let names = IndexMap::new();
-        CloneMap { tcx, names, count: 0, prelude: IndexSet::new() }
+    // TODO: When traits stop holding all functions we can remove the last two arguments
+    pub fn qname(&self, tcx: TyCtxt, def_id: DefId) -> QName {
+        self.qname_raw(method_name(tcx, def_id))
     }
 
-    pub fn name_for(&self, mut def_id: DefId, subst: SubstsRef<'tcx>) -> Option<String> {
-        if let Some(it) = self.tcx.opt_associated_item(def_id) {
-            if let ty::TraitContainer(_) = it.container {
-                def_id = it.container.id()
-            }
-        };
+    fn qname_raw(&self, method: String) -> QName {
+        QName { module: vec![self.name.clone()], name: method }
+    }
+}
 
-        Some(self.names[&(def_id, subst)].name.clone())
+impl<'tcx> CloneMap<'tcx> {
+    pub fn new(tcx: TyCtxt<'tcx>, item_type: ItemType) -> Self {
+        let names = IndexMap::new();
+        CloneMap { tcx, names, count: 0, prelude: IndexSet::new(), item_type }
     }
 
-    pub fn name_for_mut(&mut self, mut def_id: DefId, subst: SubstsRef<'tcx>) -> String {
+    pub fn insert(&mut self, mut def_id: DefId, subst: SubstsRef<'tcx>) -> &CloneInfo {
         if let Some(it) = self.tcx.opt_associated_item(def_id) {
             if let ty::TraitContainer(_) = it.container {
                 def_id = it.container.id()
             }
         };
 
-        if let Some(nm) = self.names.get(&(def_id, subst)) {
-            nm.name.clone()
-        } else {
-            let name_base = self.tcx.item_name(def_id).as_str().to_camel_case();
-            let info = CloneInfo::from_name(format!("{}{}", name_base, self.count));
-            self.names.insert((def_id, subst), info);
-            self.count += 1;
-            self.names[&(def_id, subst)].name.clone()
-        }
-    }
-
-    pub fn qname_for(&self, def_id: DefId, subst: SubstsRef<'tcx>) -> Option<QName> {
-        let module = self.name_for(def_id, subst)?;
-        Some(QName { module: vec![module], name: ctx::method_name(self.tcx, def_id) })
-    }
-
-    pub fn qname_for_mut(&mut self, def_id: DefId, subst: SubstsRef<'tcx>) -> QName {
```

Comment:
I felt like it was better to move to an `Entry` like api, rather than directly returning a `QName`, we return a reference to a `CloneInfo` from which we can build a `QName`, perhaps that is overkill now. Originally, it was meant to allow a clone to be marked as 'public' (part of the interface) or not.